### PR TITLE
AI burn score fixes and improvements

### LINKED ISSUE #5272: Fix Will-O-Wisp Scoring
### Description

Grabbing this from @iriv24 on Discord so we don't lose it:

suggestion for AI behavior: only prefer will-o-wisp if the player's mon has a physical attacking move or a higher attack stat than special attack or something along those lines.
situation where i encnoutered this:
ai has flags: AI_FLAG_CHECK_BAD_MOVE | AI_FLAG_TRY_TO_FAINT | AI_FLAG_CHECK_VIABILITY | AI_FLAG_SMART_SWITCHING | AI_FLAG_OMNISCIENT | AI_FLAG_SMART_MON_CHOICES
AI mon:
```
{
    .iv = TRAINER_PARTY_IVS(31, 31, 31, 31, 31, 31),
    .lvl = 31,
    .species = SPECIES_RAPIDASH_GALARIAN,
    .nature = NATURE_JOLLY,
    .ability = ABILITY_IMMUNITY,
    .heldItem = ITEM_SITRUS_BERRY,
    .moves = {MOVE_FLAME_WHEEL, MOVE_WILL_O_WISP, MOVE_COVET, MOVE_QUICK_ATTACK}
    },
```
Player mon: level 32 modest toedscuel with mega drain, mud shot, hex, stun spore. 
Rapidash went for will-o-wisp when i dont have any physical attack moves and have shit attack anyway (47 attack vs 79 special attack), i would expect flame wheel to be the optimal play in this scenario

### Version

1.9.1 (Latest release)

### Upcoming/master Version

_No response_

### Discord contact info

@iriv24

### PR DESCRIPTION
**Pull request by Pawkkie**
## Description
Addresses issues brought up in #5272. The ties were the result of a flat +1 score increase to any burning move as long as the AI was able to burn the target, in all cases. This increase has now been gated behind checks for the player mon having a physical attacking move that the AI knows about, or guessing that it might be a physical attacker based on its base stats. The stronger stat increase in the case of very high incoming damage was also improved to ensure as best we can that the incoming damage is physical nature before applying the score increase.

## Images
Scoring after fix
image (1)

## Issue(s) that this PR fixes
Fixes #5272 

## **People who collaborated with me in this PR**
@iriv24 and @AlexOn1ine

## **Discord contact info**
@Pawkkie @iriv24

### CODE REVIEW

File: src/battle_ai_util.c [hunk 1]
```diff
@@ -3735,11 +3735,14 @@ void IncreaseBurnScore(u32 battlerAtk, u32 battlerDef, u32 move, s32 *score)
...
     if (AI_CanBurn(battlerAtk, battlerDef, AI_DATA->abilities[battlerDef], BATTLE_PARTNER(battlerAtk), move, AI_DATA->partnerMove))
     {
-        ADJUST_SCORE_PTR(WEAK_EFFECT); // burning is good
-        if (HasMoveWithCategory(battlerDef, DAMAGE_CATEGORY_PHYSICAL))
+        if (HasMoveWithCategory(battlerDef, DAMAGE_CATEGORY_PHYSICAL)
+            || (!(AI_THINKING_STRUCT->aiFlags[battlerAtk] & AI_FLAG_OMNISCIENT) // Not Omniscient but expects physical attacker
+                && gSpeciesInfo[gBattleMons[battlerDef].species].baseAttack >= gSpeciesInfo[gBattleMons[battlerDef].species].baseSpAttack + 10))
         {
-            if (CanTargetFaintAi(battlerDef, battlerAtk))
+            if (CanTargetFaintAi(battlerDef, battlerAtk) && gMovesInfo[GetBestDmgMoveFromBattler(battlerDef, battlerAtk)].category == DAMAGE_CATEGORY_PHYSICAL)
```
Comment (AlexOn1ine) [inline src/battle_ai_util.c:3742]: is the `CanTargetFaintAi` check needed? I would say it is always useful to burn if the best damage move is physical.

Comment (Pawkkie) [reply src/battle_ai_util.c:3742]: The initial code was intending to uniquely give a score increase if the player's mon was threatening a KO pre-burn, whereas removing that check would always give the biggest score increase if the player's best move was physical, even if it only did 1 damage as opposed to 0 damage from its other moves.

Comment (AlexOn1ine) [reply src/battle_ai_util.c:3742]: Yes, but is the `CanTargetFaintAi` check useful in general? Should the condition best damage move is physical not be enough for a good increase?

Comment (Pawkkie) [reply src/battle_ai_util.c:3742]: OH I understand what you mean. Yeah I am on board with that for sure, will change :)

**Review by AlexOn1ine — APPROVED** [review]

**Comment by kittenchilly** [comment]
Oh, I believe similar stuff should've been added to the frostbite ai since it works very similarly to burn

**Comment by AlexOn1ine** [comment]
> Oh, I believe similar stuff should've been added to the frostbite ai since it works very similarly to burn

there is no frostbite equivalent to will o wisp

**Comment by kittenchilly** [comment]
> > Oh, I believe similar stuff should've been added to the frostbite ai since it works very similarly to burn
> 
> there is no frostbite equivalent to will o wisp

Oh right, it uses the same stuff as freeze so I guess not

**Comment by Pawkkie** [comment]
There's still `IncreaseFrostbiteScore` though, which for now is exclusively used for Psycho Shift. I think given that we already have that function which was a duplicate of the pre-fix `IncreaseBurnScore`, this is still a reasonable addition to make even if we don't have a specific Will O Wisp equivalent right now :)

**Pull request merged**